executor/read/mod.rs

```diff
@@ -62,35 +62,106 @@ pub(super) fn create_executors_for_pipeline(
     Ok(PatternExecutor::new(executors))
 }
 
+#[derive(Debug)]
 pub(crate) enum SuspendPoint {
     TabledCall(TabledCallSuspension),
     Nested(NestedSuspension),
 }
 
 impl SuspendPoint {
-    fn new_tabled_call(
-        executor_index: ExecutorIndex,
-        next_table_row: TableIndex,
-        input_row: MaybeOwnedRow<'static>,
-    ) -> Self {
-        Self::TabledCall(TabledCallSuspension { executor_index, next_table_row, input_row })
-    }
-
-    fn new_nested(executor_index: ExecutorIndex, branch_index: BranchIndex, input_row: MaybeOwnedRow<'static>) -> Self {
-        Self::Nested(NestedSuspension { executor_index, branch_index, input_row })
+    fn depth(&self) -> usize {
+        match self {
+            SuspendPoint::TabledCall(tabled_call) => tabled_call.depth,
+            SuspendPoint::Nested(nested) => nested.depth,
+        }
     }
 }
 
 #[derive(Debug)]
 pub(super) struct TabledCallSuspension {
     pub(crate) executor_index: ExecutorIndex,
+    pub(crate) depth: usize,
     pub(crate) input_row: MaybeOwnedRow<'static>,
     pub(crate) next_table_row: TableIndex,
 }
 
 #[derive(Debug)]
 pub(super) struct NestedSuspension {
     pub(crate) executor_index: ExecutorIndex,
+    pub(crate) depth: usize,
     pub(crate) branch_index: BranchIndex,
     pub(crate) input_row: MaybeOwnedRow<'static>,
 }
+
+#[derive(Debug)]
+pub(super) struct SuspendPointContext {
```

Comment:
I feel like `Context` is a little too generic for something that does something extremely specific - in my local branch I played with `SuspendPointCollector` since it records and pops suspend points? `Context` is hard to relate to the other data structures and algorithms at play.

Reply:
Used to be `SuspendPointTracker`